### DISCUSSION TITLE
Timeline based timing screen

### DISCUSSION THREAD

**Discussion by minetoblend** [ppy/osu#31266] [Ideas]
I want to propose replacing the current timing screen with a timeline-based ui. While the current spreadsheet-like approach works I think a timeline would bring several improvements. I've been working on a prototype of it to see how well this idea works and I like how interacting with timing points this way feels a lot so far. The prototype isn't using the lazer code base though I think I could make one for lazer fairly quickly if this proposal is deemed worth exploring. I can also provide a standalone version of the current prototype if needed before I move on to a lazer version.

<img width="1126" alt="image" src="https://github.com/user-attachments/assets/b10925b4-3113-4364-8e30-80de8c4c9924" />

Keep in mind that the types of layers present in this image don't really map one-to-one to lazer's control points (i.e. slider velocity), as it has different control point types compared to my own editor.

### Basic controls

While not present in the screenshots, I think the timeline should be paired with the current inspector window of the timing screen, which would reside next to it and remain mostly unchanged. I have some concerns with the inspector regarding selecting multiple control points, as those could be having different values. This could be solved by what other programs do often and displaying the values as "mixed", or like stable does for the control point time, which replaces it with a text field to move the selected control points by a given time span in that case.

For the basic controls I want to propose the following:

Click a control point to select it, the control key can be used for selecting multiple control points.

https://github.com/user-attachments/assets/3de97ae7-e1f8-4ac2-ab9a-20b2efcbd79e

Double click a control point to select all points at the same time.

https://github.com/user-attachments/assets/a43bd075-b111-454c-bca2-2af402d93980

Change the time of control points by dragging them in the timeline. Holding shift will disable snapping to the beat snap divisor. For timing points the snapping behavior could be inverted since you probably don't want them to snap by default (there's also concerns about the timing point snapping to itself). The inspector can be used for entering precise values.

https://github.com/user-attachments/assets/ce7a2fbd-d4d3-4c86-9cb4-bfac75b8096d

Hovering over another control point while dragging could snap to the hovered control point's time.

https://github.com/user-attachments/assets/79182b8f-68f5-4873-8ac0-313fa27d719c

### Placing control points

For placing control points I propose having 2 modes/tools which can be selected next to the timeline. When the placement tool is active, a preview of the new control point shows up and clicking anywhere on the timeline will place it. I experimented with the control key temporarily activating the placement mode too, however this would create a conflict with using ctrl for selecting multiple control points (unless maybe shift is used for that).

https://github.com/user-attachments/assets/ccbd2296-850d-4ae4-aa08-8a88f1b80adb

### Inline values

Values associated with each control point could be displayed, and editable in the timeline. They should of course still be editable through the inspector though. I'm kinda torn on having them editable in the timeline, as it tends to make them use up more space, however I do think that they should be displayed either way.

https://github.com/user-attachments/assets/82d41656-9b87-4758-ae0e-b6cfaaf2c39d

### Timing point controls

For timing points, dragging the timeline ticks could be used as a way of changing the bpm, allowing for a quick way of aligning the ticks with the waveform visually.

https://github.com/user-attachments/assets/58130546-f69d-4188-b3af-2e87a4d6a7c2

Just like when dragging a control point, hovering over another control point while dragging could also snap to the hovered control point to give a neat way of making sure the beats align between two timing points when dealing with variable bpm.

https://github.com/user-attachments/assets/a2bff93f-4295-42e7-a560-0afa31571786

### Copy-paste

Iirc you can't copy-paste timing points at all right now in lazer (kind of not feasable since you can't select multiple control points currently), however in stable you can but it's kind of confusing. With a timeline copy-pasting would be much more intuitive.

https://github.com/user-attachments/assets/82a0633d-268e-4e0f-a722-eb3215781ea3

### Kiai

I'm aware that effect points hold more than just the kiai status, but I still think it's worth looking into.
When placing the control point, dragging could set both the start & end point, allowing for a quick way of placing down a specific duration during which the state should change.

https://github.com/user-attachments/assets/ca7073d9-781e-4ecf-9f1d-af21c8ec1b95

### Remove the need for control point groups

From what I can tell from the code base, control point groups only really exist for the timing screen right now. Giving each control point type it's own timeline would allow treating them as separate instead of grouping them, and would allow removing a lot of complexity from the control point data structures.

### Two-dimensional keyframes

Keyframes that essentially just hold 1 (ranged) floating point value could use the y-axis to add a quicker way of changing the values.

https://github.com/user-attachments/assets/ccfa82af-ae00-4fce-8300-33b32d0d3d50

This is more of a toy I made just because I can but it'd be fairly simple to expand this to stuff like easing functions or even cubic-bezier easing. Could be useful if a timeline like this were used for storyboarding though.

https://github.com/user-attachments/assets/520af6d8-91e3-422a-9faa-eb32336f11aa

**Comment by minisbett** [comment]
This would be an insane addition, I love this idea so much

**Comment by peppy** [comment]
This was part of my design proposal to @arflyte originally, which has been partly realised in some of the editor designs. Which is to say, I'm all for this direction if implemented correctly.

Note: I haven't read over the specifics here, just the basic idea. If someone is to take this on, it should be discussed further with myself.

**Reply by minetoblend** [reply]
I'd be happy to implement this if given the chance. I've been working on a small prototype over the days since I wrote this proposal to demonstrate the basic functionality (not quite finished yet though). Could this work as a starting point for further discussion?

https://github.com/user-attachments/assets/783c2ab2-1e42-4ce9-b125-df90e850fb18

**Reply by peppy** [reply]
I dunno I'd see it being on the timing screen like you have it. Also dragging timing points in place like that doesn't look precise. Also effect points aren't snapping (and you don't have the beat snapping options visible to give the user a choice of whether they want it to snap).

Where is the control point list going to go? It should still be visible.

How would I see this working? I believe it should be extending the *timeline* area, not implemented as its own stand-alone system as you have it. The rest of the editor should still be accessible and show as it was – the timeline would just extend further vertically.

See https://www.figma.com/design/ytnnne2TH8Z956Jxhiypqq/Beatmap-Editor-2?node-id=0-1&p=f for some initial ideas. The timeline can expand down as much as 50% of the screen IMO.

**Reply by minetoblend** [reply]
Agree, the list should still be accessible. Making the list work with multi-select was a bit out of scope for what I wanted to show with the prototype, so I removed it for that.

I was originally thinking of having there be a toggle to switch between a list mode and a timeline mode, though having them stacked like in the figma design does look better to me (especially since there's only 2 types of control points currently, so not that much vertical space needed anyways).

> Also dragging timing points in place like that doesn't look precise.

I'm imagining dragging the timing points to mainly be of use for songs with variable bpm, where the ease of quickly making edits and being able to visually align the points with the waveform should make for a good results quicker. Timing songs like that is currently very cumbersome and can take several hours.
I think that for adjusting the bpm by dragging, having an inline popup appear afterwards that lets you enter a precise value (and perhaps showing a button for applying the bpm rounded to the nearest number) should also help with precision. 
With the inspector to the right it's also still possible to just enter the precise values.

> Also effect points aren't snapping (and you don't have the beat snapping options visible to give the user a choice of whether they want it to snap).

That's on me, I was zoomed out a bit too far on the timeline so it didn't really show. The way I have it right now it will snap by default (except for timing points) & snapping can be temporarily disabling it by pressing shift while dragging. Having a dedicated place for it in the ui also doesn't sound bad though.
Control points can also be snapped to other points by dragging on top of another control point. That is especially useful for aligning the white tick of the previous timing point with the one afterwards when adjusting a timing point on a song with variable bpm.

https://github.com/user-attachments/assets/e5457927-49cd-4c31-bed3-369a20707477

**Reply by peppy** [reply]
For all timing facilities, I'd probably see them being in the same timing control that is to the right. Or at least tightly integrated with it. I'd rather try for that before creating two different timing experiences that have completely different workflows.

Also I'd see this system providing hitsounding possibilities as well, via extensions to the timeline (see https://github.com/ppy/osu/pull/31291 which is a bit of a mess but shows the user case for it).

**Reply by minetoblend** [reply]
I'm not sure how the timing tools in the inspector could be turned into something that is really usable for variable bpm, timing those kinds of songs has very different requirements compared to constant bpm and I think trying to force a single workflow for both is not a step in the right direction.

Being able to adjust timing points on the timeline like this is the main reason why I'm pushing for an extended timeline in the first place,  without that I'm not sure how what the benefit of having it would be.

**Reply by peppy** [reply]
I guess see how you go without spending too much time and get a proof-of-concept I can play with. I'd want to be convinced that this is something I can agree with in the editor before getting too deep in discussions.

**Reply by minetoblend** [reply]
Sounds good. Should I open a draft pr once I got it in a good state or would you prefer I just link the branch here?

**Reply by peppy** [reply]
Draft PR is fine 👍

**Reply by velamyGH** [reply]
hello please do not let this die i would love to see this come back into discussion :pray:
